### ISSUE TITLE
LoRA Fine-Tuning on LIBERO: 0% Success Rate Despite Training Loss Convergence

### ISSUE THREAD

**Issue by kakaxide** [RLinf/RLinf#1557]
### Summary

When fine-tuning the `pi05_libero_base` model with LoRA (rank=32) on LIBERO tasks using RLinf's SFT pipeline, training loss converges to ~0.001, but **all evaluations yield 0% success rate**. In contrast, a fully fine-tuned model (`pi05_libero_finetuned_v044`, trained outside RLinf) achieves **18%–84% success rate** on the same evaluation setup. This issue persists across multiple training runs with different hyperparameters.

### Environment

- **Base model**: `lerobot/pi05_libero_base`
- **LoRA rank**: 32
- **Dataset**: LIBERO object (10 tasks) / LIBERO-10 task0
- **Training framework**: RLinf SFT with FSDP
- **Evaluation**: 48–50 trajectories, 280–520 max steps per episode

### Training Configuration

```yaml
runner:
  max_steps: 6500
  save_interval: 1000

data:
  train_data_paths: "/private/RLinf/datasets/libero_object"

actor:
  micro_batch_size: 7
  global_batch_size: 28
  model:
    model_path: "lerobot/pi05_libero_base"
    num_action_chunks: 4
    is_lora: True
    lora_rank: 32
    openpi:
      config_name: "pi05_libero"
  optim:
    lr: 2e-4
    lr_scheduler: "cosine"
    lr_warmup_steps: 650
    total_training_steps: 6500
    min_lr: 2e-5
```

### Training Results — All Runs Converge

| Run | LR | Steps | Batch Size | Init Loss | Final Loss |
|-----|----|-------|------------|-----------|------------|
| 1   | 2.5e-5 | 10000 | 28 | ~0.5 | ~0.001 |
| 2   | 2.5e-5 | 2000  | 56 | ~0.536 | ~0.002 |
| 3   | 1e-4   | 6500  | 28 | ~0.499 | ~0.001 |
| 4   | 2.5e-5 | 10000 | 48 | ~0.5   | ~0.001 |

All runs show smooth loss convergence.

#### RLinf LoRA Fine-Tuned Checkpoints (all 0% success)

| Eval Date | Checkpoint | num_action_chunks | Success Rate | Trajectories |
|-----------|-----------|-------------------|-------------|-------------|
| Sep 10 | global_step_10000 (Run 1) | 5 | **0%** | 48 |
| Sep 10 | global_step_1000 (Run 1)  | 5 | **0%** | 48 |
| Sep 10 | global_step_2000 (Run 2)  | 5 | **0%** | 48 |
| Sep 11 | global_step_6500 (Run 3)  | 4 | **0%** | 48 |
| Sep 11 | global_step_6500 (Run 3)  | 4 | **0%** | 48 |
| Sep 6  | global_step_8000 (Run 4)  | 4 | **0%** | 50 |

#### External Fully Fine-Tuned Model (`pi05_libero_finetuned_v044`)

| Eval Date | Task Setup | num_action_chunks | Success Rate | Trajectories |
|-----------|-----------|-------------------|-------------|-------------|
| Sep 7  | libero_object | 5 | **84%**  | 50 |
| Sep 10 | libero_object | 5 | **79.2%** | 48 |
| Sep 7  | libero_10_task0 | 4 | **34%** | 50 |
| Sep 7  | libero_10 | 4 | **18%** | 50 |

#### Other External SFT Model (`RLinf-Pi05-LIBERO-SFT`)

| Eval Date | Task Setup | num_action_chunks | Success Rate |
|-----------|-----------|-------------------|-------------|
| Sep 6 | libero_10_task0 | 4 | **0%** |

### Key Observations

1. **All RLinf LoRA SFT checkpoints achieve 0% success**, regardless of learning rate (2.5e-5 or 1e-4), training steps (2k–10k), batch size (28–56), or `num_action_chunks` (4 or 5).

2. **The externally fine-tuned `v044` model achieves 18%–84% success** on the same evaluation pipeline, depending on task setup and `num_action_chunks`.

3. **Training loss converges normally** (~0.001) in all cases, suggesting the model fits the training data.

4. **LoRA weight loading is verified correct**: `[openpi load] mapped 813 tensors (lora pairs merged: 289); missing=0 unexpected=0` — 289 LoRA pairs successfully merged into base weights.

5. **Learning rate is not the differentiating factor**: both 2.5e-5 and 1e-4 produce similar loss curves and both yield 0% evaluation success.

### Potential Issue: Gripper Action Convention in `prepare_actions_for_libero`

In `rlinf/envs/action_utils.py`, the `prepare_actions_for_libero` function applies gripper action conversion only for specific model types:

```python
def prepare_actions_for_libero(raw_chunk_actions, model_type) -> np.ndarray:
    chunk_actions = raw_chunk_actions
    if SupportedModel(model_type) in [
        SupportedModel.OPENVLA,
        SupportedModel.OPENVLA_OFT,
        SupportedModel.GR00T_N1D6,
        SupportedModel.GR00T_N1D7,
    ]:  # SupportedModel.OPENPI / OPENPI_PYTORCH not included
        chunk_actions[..., -1] = 2 * chunk_actions[..., -1] - 1
        chunk_actions[..., -1] = np.sign(chunk_actions[..., -1]) * -1.0
    return chunk_actions
```

`SupportedModel.OPENPI` is **not** in the conversion list. The LIBERO dataset uses gripper ∈ [0, 1] (1=open, 0=close), while the `PandaGripper` environment expects +1=close, -1=open. Without conversion, model outputs in [0, 1] are passed directly to the environment.

However, the `v044` model also goes through this same code path without conversion and achieves 79%+ success. This raises the question of whether the convention mismatch is the actual root cause, or if there is a difference in how the LoRA SFT pipeline processes actions compared to the external training pipeline.

### Questions for Maintainers

1. Why do all RLinf LoRA SFT checkpoints achieve 0% success while the externally fine-tuned `v044` model works on the same evaluation pipeline? Is there a known difference in action representation or normalization between models trained via RLinf SFT vs. external training?

2. Should `SupportedModel.OPENPI` be included in the gripper conversion list in `prepare_actions_for_libero`? If not, why does the `v044` model work without it — does it output gripper values in a different range?

3. Are there any known issues with LoRA fine-tuning for openpi models in RLinf beyond the action post-processing? Could the LoRA adapter configuration (rank=32, target modules) or the FSDP checkpoint saving/loading introduce subtle issues?

4. The `RLinf-Pi05-LIBERO-SFT` model (an external SFT checkpoint) also achieves 0% on RLinf's evaluation pipeline — is this expected, or does it indicate a broader evaluation pipeline issue?

5. Could `num_action_chunks` mismatch between training and evaluation affect results? The LoRA models were trained with `num_action_chunks=4` but some evaluations used `num_action_chunks=5` (and vice versa). The `v044` model achieves its best result (84%) with `num_action_chunks=5`.

**Comment by zoeyuchao** [comment]
@kakaxide have you used openpi rlinf version？

**Comment by kakaxide** [comment]
“Yes, I’ve used the OpenPI RLinf version.
